gradescopeCronJob/gradescope_to_spreadsheet.py

```diff
@@ -110,14 +112,25 @@ def get_sub_sheet_titles_to_ids(sheet_api_instance):
     if subsheet_titles_to_ids:
         return subsheet_titles_to_ids
     print("Retrieving subsheet titles to ids")
-    sheets = sheet_api_instance.get(spreadsheetId=SPREADSHEET_ID, fields='sheets/properties').execute()
+    request = sheet_api_instance.get(spreadsheetId=SPREADSHEET_ID, fields='sheets/properties')
+    sheets = request.execute()
     subsheet_titles_to_ids = {sheet['properties']['title']: sheet['properties']['sheetId'] for sheet in
                                sheets['sheets']}
     return subsheet_titles_to_ids
 
+def backoff_handler():
+    pass
+@backoff.on_exception(
+    backoff.expo,
+    Exception,
+    max_tries=sys.maxsize,
```

Comment:
Think about edge cases. What if someone accidentally deletes this spreadsheet? Then, we have an infinite loop in this program because max_tries = `sys.maxsize`, which essentially means try an infinite number of times.

Assumptions: The script will run every X = 5 minutes.
Note: If the cron job execution 1 is not finished, cron job execution 2 will still start. If the cron job takes 6 minutes to run and starts at time t = 0 minutes, cron job 2 will still start at time = 5 minutes.
Claim: This script runs in 1 minute.

Suppose we change this to a number like `max_size=10`. 

Upon the first failure, if the `factor = 1`, we wait pow(2,0) * `factor` = 1 second
Upon the second failure, we wait pow(2,1) = 2 seconds before retrying
... Upon the 10th failure, we wait pow(2,9) = 512 seconds
`2**10 - 1 = sum([pow(2, i) for i in range(0, 10)])` = 1023 seconds = 17.05 minutes. 

This is too long to wait because 3 cron jobs will start executing while the first cron job is still executing. Let's try again.

Suppose we instead change this to be `max_size = 5`, then `2**6 - 1 = sum([pow(2, i) for i in range(0, 6)])` = 63 seconds < X = 300 seconds = 5 minutes.

This is already pretty high. Based on your experience, please change max_size to be a number such that 2**`max_size+1` - 1 <= (frequency the cron job will occur in seconds). My suggestion is `max_size = 5` if we do `X = 5 minutes`.

Reply:
This is a really good intuition.  I personally think the optimal approach is to prevent the requeue on cron execution if the prior execution has not yet terminated.  I like this approach for two reasons:

1. We already are handling exponential, error-state backoff requeuing which is the proper way to requeue jobs dependent on inconsistent operations.
2. The idea of this job is to flush the local data to the db (sheets).  If we have multiple parallel execution runtimes, each runtime will have its own distinct memory allocation to process the data.  This is an issues as demonstrated: 

    Let $S_t$ be hosts local persistent storage at timestep $t$ and $J_t$ be the job execution environment triggered at timestep $t$.  If $J_t$ loads the current state of the db to its allocated memory at timestep $t$, then we can say that __lemma 1__: the memory state at time $t$ for job $J_t$ denoted as $M(j_t)$ is equal to $S_t$.

    Assume for contradiction that, at timestep $t$, the state of the database $D$, (google sheets) is consistent with $S_t$ denoted by $\forall t, r(D_t) = r(S_t)$.  Assume wlog we have two execution environments $J_0$ and $J_1$ and $S_0 \neq S_1$.  It follows by _lemma 1_ that $M(J_0) = S_0$ and $M(J_1) = S_1$.  It proceeds by cases that:

    1. $w(S)D$ is atomic:
        1. $w(M(J)D)$ may fail $\implies \exists \\{ \\{w(M(J_i))D_a, w(M(J_k))D_b | k \gt i, a \gt b\\}, \tau \in \mathbb{Z}^+ \\} \ni D_{b+\tau} = D_a = S_i$
        1. $w(M(J_t))D_a \implies t \leq a \therefore \exists b > a \ni \\{w(M(J_i))D_a, w(M(J_k))D_b | i > k\\} \neq \emptyset$
     1. $w(S)D$ is not atomic is trivially susceptible to race conditions but is likewise susceptible to the atomic counterexamples above.

    $\blacksquare$

Reply:
Thanks a ton for the thought you put into this, @Mehul-Gandhi  and @Connor-Bernard!!

@Mehul-Gandhi, will set max_size to 5!

@Connor-Bernard, totally makes sense to avoid beginning the next cron execution before the current one ends!